File: 2. LeetCode/238. Product of Array Except Self.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int> &nums) {
        vector<int> left(nums.size());
        vector<int> right(nums.size());
        vector<int> res(nums.size());

        for (int i = 1; i < nums.size(); ++i)
            left[i] = left[i - 1] * nums[i - 1];
        for (int i = nums.size() - 2; i >= 0; --i)
            right[i] = right[i + 1] * nums[i + 1];

        for (int i = 0; i < nums.size(); ++i)
            res[i] = left[i] * right[i];

        return res;
    }
};

File: LeetCode/238. Product of Array Except Self.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        left = [1] * n
        right = [1] * n
        res = [1] * n

        # 왼쪽 누적 곱 계산
        for i in range(1, n):
            left[i] = left[i - 1] * nums[i - 1]

        # 오른쪽 누적 곱 계산
        for i in range(n - 2, -1, -1):
            right[i] = right[i + 1] * nums[i + 1]

        # 결과 리스트 계산
        for i in range(n):
            res[i] = left[i] * right[i]

        return res


s = Solution()
print(s.productExceptSelf(nums=[1, 2, 3, 4]))
